### ISSUE TITLE
Shared object build fails on CentOS7 when building with DPDK master and SPDK 19.07.x

### ISSUE THREAD

**Issue by karlatec** [spdk/spdk#960]
## Current Behavior
Nightly tests spdk-19.07-vs-dpdk-master builds fail.
The reason is failure on CentOS7 machines when building shared objects.

## Steps to Reproduce
1. Clone dpdk from http://dpdk.org/git/dpdk, check-out master branch (can use specific commit from Context section), build it with 
```
make config T=x86_64-native-linuxapp-gcc
```
2.  Use SPDK commit from context section below. Build with options:
```
SPDK_BUILD_DOC=0
SPDK_BUILD_IOAT_KMOD=0
SPDK_BUILD_SHARED_OBJECT=1
SPDK_FEDORA_PACKAGING=0
SPDK_RUN_CHECK_FORMAT=1
SPDK_RUN_SCANBUILD=0
SPDK_RUN_VALGRIND=0
SPDK_RUN_FUNCTIONAL_TEST=1
SPDK_TEST_UNITTEST=1
SPDK_TEST_ISCSI=0
SPDK_TEST_NVME=0
SPDK_TEST_BDEV_OCSSD=0
SPDK_TEST_FTL=1
SPDK_TEST_NVMF=0
SPDK_TEST_RBD=0
SPDK_TEST_VHOST=0
SPDK_TEST_BLOCKDEV=1
SPDK_TEST_IOAT=0
SPDK_TEST_EVENT=1
SPDK_TEST_BLOBFS=0
SPDK_TEST_NVML=0
SPDK_TEST_LVOL=0
SPDK_RUN_VPP=0
SPDK_TEST_CRYPTO=0
SPDK_TEST_REDUCE=0
SPDK_TEST_OCF=0
SPDK_TEST_ISAL=0
SPDK_RUN_ASAN=1
SPDK_RUN_UBSAN=0
SPDK_TEST_VHOST_INIT=0
RUN_NIGHTLY=0
SPDK_TEST_CRYPTO=0


./configure --enable-debug --enable-werror --enable-log-bt \
--with-rdma --with-fio=/usr/src/fio --enable-asan --enable-coverage \
--without-isal --with-dpdk=/var/jenkins/workspace/Other_systems/centos7_autotest_nightly/dpdk/build \
--with-shared
```
3. Result:
```
[...]
---------------------------------------------------------------------
There are still undefined symbols in the library ut_mock 
__real_calloc
__real_pthread_mutexattr_
__real_pthread_mutex_init
---------------------------------------------------------------------
There are still undefined symbols in the library rte_vhost 
get_mempolicy
rte_free
rte_log
rte_malloc
rte_malloc_socket
rte_mempool_ops_table
rte_mem_virt2phy
__rte_panic
rte_zmalloc
---------------------------------------------------------------------
there was a dependency mismatch in the library vhost 
The makefile lists: 'bdev conf event json jsonrpc log rpc scsi thread util'
readelf outputs   : 'bdev conf event json jsonrpc log rpc rte_vhost scsi thread util'
---------------------------------------------------------------------
shared object test failed
```

## Context (Environment including OS version, SPDK version, etc.)
DPDK: bd253da devtools: fix test of ninja install
SPDK: a2f6965f2 branch v19.07.x

**Comment by karlatec** [comment]
Related fails:
https://dqtibwqq6s6ux.cloudfront.net/public_build/autotest-spdk-19-07-vs-dpdk-master-chandler_304.html
https://dqtibwqq6s6ux.cloudfront.net/public_build/autotest-spdk-19-07-vs-dpdk-master-chandler_303.html
https://dqtibwqq6s6ux.cloudfront.net/public_build/autotest-spdk-19-07-vs-dpdk-master_220.html
https://dqtibwqq6s6ux.cloudfront.net/public_build/autotest-spdk-19-07-vs-dpdk-master_219.html

**Comment by Seth5141** [comment]
Hi Karol,

This failure is caused by a couple of things.
1. The proper dependencies haven't been specified for the internal rte_vhost implementation on v19.07.x. There is a patch out for rebase which fixes this. It has already been merged to master, so that's why we don't run into it on the spdk-master-vs-dpdk-master job. Here is the link to that patch. https://review.gerrithub.io/c/spdk/spdk/+/467947. Once this change is backported, the failure will go away, but there is still a latent problem in the configure script (see 2).
2. The CentOS machine shouldn't be tryng to build the internal rte_vhost library anyways since the version of DPDK we are checking against supports the features we need. If you look at any of the logs you posted, you will find the following lines:
~~~
 06:31:58	$ ./configure --enable-debug --enable-werror --enable-log-bt --with-rdma --with-fio=/usr/src/fio --enable-asan --enable-coverage --without-isal --with-dpdk=/var/jenkins/workspace/Other_systems/centos7_autotest_nightly/dpdk/build
Using default SPDK env in /var/jenkins/workspace/Other_systems/centos7_autotest_nightly/spdk/lib/env_dpdk
Notice: DPDK's rte_vhost not found or version < 19.05, using internal, legacy rte_vhost library.
~~~
Obviously the version of DPDK we are testing against is > 19.05, but the check in configure is still failing. I ran into a similar problem on my local dev machine before and thought I had this fixed (see https://review.gerrithub.io/c/spdk/spdk/+/467714), but it turns out there is a different problem on CentOS7. When we try to run the configure test as is, it fails with a bunch of type conversion errors.
When I add the -march=native flag to that build, then it works properly.
I'll push a change with the right flags in BUILD_CMD and verify that CentOS picks up the change properly.

**Comment by Seth5141** [comment]
This change fixes part 2. https://review.gerrithub.io/c/spdk/spdk/+/468636.
Please see the logs at
https://dqtibwqq6s6ux.cloudfront.net/results/autotest-spdk-master-vs-dpdk-master/builds/288/archive/centos7_autotest/build.log

and 

https://dqtibwqq6s6ux.cloudfront.net/results/autotest-spdk-master-vs-dpdk-master/builds/287/archive/centos7_autotest/build.log

These are spdk master vs dpdk  master logs, but they demonstrate the problem. The first log is run against the change I linked above, and the second one is against master. The second log contains the line
~~~
Notice: DPDK's rte_vhost not found or version < 19.05, using internal, legacy rte_vhost library.
~~~

while the first one doesn't.

tl;dr once https://review.gerrithub.io/c/spdk/spdk/+/467947 is backported, and https://review.gerrithub.io/c/spdk/spdk/+/468636 is merged (doesn't necessarily need a backport since we will just change the 19.07 job to 19.10 soon and the change isn't critical), I think we can close this.

**Comment by Seth5141** [comment]
closing. The patches mentioned have been merged/backported and the nightly spdk19.07 v dpdk-master build is passing. https://dqtibwqq6s6ux.cloudfront.net/public_build/autotest-spdk-19-07-vs-dpdk-master_346.html
